From f014e09434ee006e96ea5a5e935122ca8ef3c55f Mon Sep 17 00:00:00 2001
From: Michal Orzel <michalx.orzel@intel.com>
Date: Mon, 7 Nov 2022 15:17:19 +0100
Subject: [PATCH] curl: Fix verification of CURLOPT_TLS13_CIPHERS

The code checking for CURLOPT_TLS13_CIPHERS option did not work
properly, because of incorrect assumption that this symbol was a
preprocessor macro. It is in fact element of enum type, which
resulted with #ifdef directive working improperly. Change replaces
compile-time verification with run-time, based on return value of
curl_easy_setopt function.

Tested:
Nbdkit now works properly and doesn't exit prematurely.

Signed-off-by: Michal Orzel <michalx.orzel@intel.com>
---
 plugins/curl/curl.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/plugins/curl/curl.c b/plugins/curl/curl.c
index 77f88fff..4fb62616 100644
--- a/plugins/curl/curl.c
+++ b/plugins/curl/curl.c
@@ -560,14 +560,13 @@ curl_open (int readonly)
   if (ssl_cipher_list)
     curl_easy_setopt (h->c, CURLOPT_SSL_CIPHER_LIST, ssl_cipher_list);
   if (tls13_ciphers) {
-#ifdef CURLOPT_TLS13_CIPHERS
-    curl_easy_setopt (h->c, CURLOPT_TLS13_CIPHERS, tls13_ciphers);
-#else
-    /* This is not available in, eg, RHEL 7 */
-    nbdkit_error ("tls13-ciphers is not supported in this build of "
-                  "nbdkit-curl-plugin");
-    goto err;
-#endif
+    r = curl_easy_setopt (h->c, CURLOPT_TLS13_CIPHERS, tls13_ciphers);
+    if (r != CURLE_OK) {
+      /* This is not available in, eg, RHEL 7 */
+      display_curl_error (h, r, "curl_easy_setopt: CURLOPT_TLS13_CIPHERS [%s]",
+        tls13_ciphers);
+      goto err;
+    }
   }
   if (tcp_keepalive)
     curl_easy_setopt (h->c, CURLOPT_TCP_KEEPALIVE, 1L);
-- 
2.25.1

